pkgs/top-level/config.nix

```diff
@@ -0,0 +1,199 @@
+# This file defines the structure of the `config` nixpkgs option.
+
+{ lib, localSystem, crossSystem, config, ... }:
+
+with lib;
+
+let
+
+  mkMeta = args: mkOption (builtins.removeAttrs args [ "feature" ] // {
+    type = types.uniq (args.type or types.bool);
+    default = args.default or false;
+    description = args.description or ''
+      Whether to ${args.feature} while evaluating nixpkgs.
+    '' + ''
+      Changing the default will not cause any rebuilds.
+    '';
+  });
+
+  mkMassRebuild = args: mkOption (builtins.removeAttrs args [ "feature" ] // {
+    type = types.uniq (args.type or types.bool);
+    default = args.default or false;
+    description = (args.description or ''
+      Whether to ${args.feature} while building nixpkgs packages.
+    '') + ''
+      Changing the default will cause a mass rebuild.
+    '';
+  });
+
+  mkUse = args: mkOption (builtins.removeAttrs args [ "feature" ] // {
+    type = types.uniq (args.type or types.bool);
+    default = args.default or false;
+    description = (args.description or ''
+      Whether to enable ${args.feature} support for nixpkgs packages.
+    '') + ''
+      Changing the default will cause a mass rebuild.
+    '';
+
+    # Note: this can't be replaced with `apply` because we want to
+    # keep the value as-is for references via `config.use`.
+    extra = args.extra or (x: {});
+  });
+
+  whenEnabled = args: value: lib.optionalAttrs value args;
+  whenDisabled = args: value: lib.optionalAttrs (!value) args;
```

Comment:
These are not used anywhere, right? Maybe add a comment explaining the argument order swap?

Reply:
Oops, sorry, only `whenEnabled` is not used — and I see the currying benefits of `whenDisable` — which might still be worth a comment.